src/ATokenNaive.sol

```diff
@@ -0,0 +1,357 @@
+import {WadRayMath} from "aave-v3-core/contracts/protocol/libraries/math/WadRayMath.sol";
+import {SafeCast} from "openzeppelin-contracts/contracts/utils/math/SafeCast.sol";
+import {Math} from "openzeppelin-contracts/contracts/utils/math/Math.sol";
+
+interface IFractionalGovernor {
+  function token() external returns (address);
+  function proposalSnapshot(uint256 proposalId) external returns (uint256);
+  function proposalDeadline(uint256 proposalId) external view returns (uint256);
+  function castVoteWithReasonAndParams(
+    uint256 proposalId,
+    uint8 support,
+    string calldata reason,
+    bytes memory params
+  ) external returns (uint256);
+}
+
+interface IVotingToken {
+  function transfer(address to, uint256 amount) external returns (bool);
+  function transferFrom(address from, address to, uint256 amount) external returns (bool);
+  function delegate(address delegatee) external;
+  function getPastVotes(address account, uint256 blockNumber) external returns (uint256);
+}
+
+contract ATokenNaive is AToken {
+  using WadRayMath for uint256;
+  using SafeCast for uint256;
+  using GPv2SafeERC20 for IERC20;
+
+  /// @notice The voting options corresponding to those used in the Governor.
+  enum VoteType {
+    Against,
+    For,
+    Abstain
+  }
+
+  /// @notice Data structure to store vote preferences expressed by depositors.
+  struct ProposalVote {
+    uint128 againstVotes;
+    uint128 forVotes;
+    uint128 abstainVotes;
+  }
+
+  /// @notice Must call castVote within this many blocks of the proposal
+  /// deadline, so-as to allow ample time for all depositors to express their
+  /// vote preferences. Corresponds to roughly 4 hours, assuming 12s/block.
+  uint32 public constant CAST_VOTE_WINDOW = 1200;
+
+  /// @notice Map depositor to deposit amount.
+  mapping(address => uint256) public deposits;
+
+  /// @notice Map borrower to total amount borrowed.
+  mapping(address => uint256) public borrowTotal;
+
+  /// @notice Map proposalId to an address to whether they have voted on this proposal.
+  mapping(uint256 => mapping(address => bool)) private _proposalVotersHasVoted;
+
+  /// @notice Map proposalId to vote totals expressed on this proposal.
+  mapping(uint256 => ProposalVote) public proposalVotes;
+
+  /// @notice The governor contract associated with this governance token.
+  IFractionalGovernor public immutable governor;
+
+  /// @dev Constructor.
+  /// @param _pool The address of the Pool contract
+  /// @param _governor The address of the flex-voting-compatable governance contract.
+  constructor(IPool _pool, address _governor) AToken(_pool) {
+    governor = IFractionalGovernor(_governor);
+  }
+
+  // TODO Is there a better way to do this? It cannot be done in the constructor
+  // because the AToken is just used a proxy -- it won't share an address with
+  // the implementation (i.e. this code).
+  function selfDelegate() public {
+    IVotingToken(governor.token()).delegate(address(this));
+  }
```

Comment:
Open to suggestions on this

Reply:
> can this logic be in that proxy's initializer method that it calls?

It could be, we'd just need to make that initializer virtual -- [currently it's not](https://github.com/aave/aave-v3-core/blob/f3e037b3638e3b7c98f0c09c56c5efde54f7c5d2/contracts/protocol/tokenization/AToken.sol#L53-L62) -- and then override it in our implementation. Might be worth asking -- I bet they'd let us

> Often you pass arbitrary data to delegatecall on the new implementation to initialize it

Initialization is done by the Aave `PoolConfigurator` [here](https://github.com/aave/aave-v3-core/blob/f3e037b3638e3b7c98f0c09c56c5efde54f7c5d2/contracts/protocol/libraries/logic/ConfiguratorLogic.sol#L245), right after the proxy is deployed. So our hands are tied there.

There is a bit of arbitrary data that we can pass into the initiailzer function, specifically the `bytes params` [argument](https://github.com/aave/aave-v3-core/blob/f3e037b3638e3b7c98f0c09c56c5efde54f7c5d2/contracts/protocol/libraries/types/ConfiguratorInputTypes.sol#L20). The issue is just that the initializer, as written, [does nothing](https://github.com/aave/aave-v3-core/blob/f3e037b3638e3b7c98f0c09c56c5efde54f7c5d2/contracts/protocol/tokenization/AToken.sol#L53-L84) with those bytes.

Reply:
Actually I think we'd just want to add a new initializer, because we don't want to re-initialize the AToken. Then to actually do the upgrade you'd use this [`upgradeToAndCall`](https://github.com/aave/aave-v3-core/blob/f3e037b3638e3b7c98f0c09c56c5efde54f7c5d2/contracts/protocol/libraries/aave-upgradeability/BaseImmutableAdminUpgradeabilityProxy.sol#L69) method to pass in the new AToken implementation and pass the `data` to delegatecall, which for us might just be calling `selfDelegate`, so we don't necessarily need a new initialization method (and maybe we just ensure that can only be called, to be safe, by reverting if we're already self delegating)

There may be a function they have to bulk upgrade all AToken implementations, but in general that's the method you use to ugprade these type of proxies. (I just went to the [aUSDC token](https://optimistic.etherscan.io/address/0x625E7708f30cA75bfd92586e17077590C60eb4cD#writeContract) on optimistic etherscan and clicked write contract to find that method, since I wasn't sure what proxy pattern they use)

Reply:
Ohhhh, I see what you mean now. That's actually really nice!

Now that I'm thinking about it, I wonder if the test setup here should actually be:
* deploy a traditional aToken for GOV
* upgrade the aToken to the ATokenNaive implementation (self-delegating in the process)
* run test

My only fear about changing the implementation is that we add some storage slots with our aToken implementation. I _think_ those will just get tacked on at the end (since we're inheriting from their AToken), but I should confirm this. If they mess up storage locations that would be _bad_

Reply:
Agree with both using that test setup and verifying we don't shift storage slots !(which, I agree we shouldn't if we inherit from AToken, but verifying that + changing the test setup can be a separate issue/PR)

Reply:
Tracking here https://github.com/ScopeLift/flexible-voting/issues/14